python/pyspark/sql/connect/client/reattach.py

```diff
@@ -0,0 +1,166 @@
+                        try:
+                            self._current = next(self._iterator)
+                        except StopIteration:
+                            pass
+                    has_next = self._current is not None
+
+                    # Graceful reattach:
+                    # If iterator ended, but there was no ResponseComplete, it means that
+                    # there is more, and we need to reattach. While ResponseComplete didn't
+                    # arrive, we keep reattaching.
+                    if not has_next and not self._response_complete:
+                        while not has_next:
+                            self._iterator = self._stub.ReattachExecute(
+                                self._create_reattach_execute_request()
+                            )
+                            # shouldn't change
+                            assert not self._response_complete
+                            try:
+                                self._current = next(self._iterator)
+                            except StopIteration:
+                                pass
+                            has_next = self._current is not None
+                    return has_next
+            return False
+
+    def _release_execute(self, until_response_id: Optional[str]) -> None:
+        request = self._create_release_execute_request(until_response_id)
+        for attempt in Retrying(can_retry=SparkConnectClient.retry_exception, **self._retry_policy):
+            with attempt:
+                self._stub.ReleaseExecute(request)
+
+    def _create_reattach_execute_request(self) -> pb2.ReattachExecuteRequest:
+        release = pb2.ReattachExecuteRequest(
+            session_id=self._initial_request.session_id,
+            user_context=self._initial_request.user_context,
+            operation_id=self._initial_request.operation_id,
+        )
+
+        if self._initial_request.client_type:
+            release.client_type = self._initial_request.client_type
+
+        return release
+
+    def _create_release_execute_request(
+        self, until_response_id: Optional[str]
+    ) -> pb2.ReattachExecuteRequest:
+        release = pb2.ReattachExecuteRequest(
+            session_id=self._initial_request.session_id,
+            user_context=self._initial_request.user_context,
+            operation_id=self._initial_request.operation_id,
+        )
+
+        if self._initial_request.client_type:
+            release.client_type = self._initial_request.client_type
+
+        if not until_response_id:
+            release.release_type = (  # type: ignore[attr-defined]
+                pb2.ReleaseExecuteRequest.ReleaseType.RELEASE_UNTIL_RESPONSE
+            )
+        else:
+            release.release_type = (  # type: ignore[attr-defined]
+                pb2.ReleaseExecuteRequest.ReleaseType.RELEASE_ALL
+            )
+            release.until_response_id = until_response_id  # type: ignore[attr-defined]
```

Comment:
heads up: changed to not have release_type enum + a oneof, but just a single oneof.